binary/test/Test/Pos/Binary/BiSerialize.hs

```diff
@@ -0,0 +1,152 @@
+{-# LANGUAGE DeriveAnyClass  #-}
+{-# LANGUAGE TemplateHaskell #-}
+module Test.Pos.Binary.BiSerialize
+    ( tests
+    ) where
+
+import           Universum
+
+import           Hedgehog (Gen, Property, discover, (===))
+import qualified Hedgehog as H
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+
+import           Pos.Binary.Class (Cons (..), Field (..), cborError, deriveIndexedBi,
+                                   deriveSimpleBi, serialize')
+
+import qualified Serokell.Util.Base16 as B16
+
+
+--------------------------------------------------------------------------------
+
+data TestSimple
+    = TsInt
+        { unTsInt :: Int }
+    | TsIntList
+        { unTsIntList :: [Int] }
+    | TsChar2
+        { unTsChar2L :: Char
+        , unTsChar2R :: Char }
+    | TsInteger
+        { unTsInteger :: Integer }
+    | TsMaybeInt
+        { unTsMaybeInt :: Maybe Int }
+    | TsChar2Permuted
+        { unTsChar2PermutedL :: Char
+        , unTsChar2PermutedR :: Char }
+    | TsPair
+        { unTsPairL :: TestSimple
+        , unTsPairR :: TestSimple }
+    deriving (Eq, Show, Typeable)
+
+deriveSimpleBi ''TestSimple [
+    Cons 'TsInt [
+        Field [| unTsInt            :: Int        |]
+        ],
+    Cons 'TsIntList [
+        Field [| unTsIntList        :: [Int]      |]
+        ],
+    Cons 'TsChar2 [
+        Field [| unTsChar2L         :: Char       |],
+        Field [| unTsChar2R         :: Char       |]
+        ],
+    Cons 'TsInteger [
+        Field [| unTsInteger        :: Integer    |]
+        ],
+    Cons 'TsMaybeInt [
+        Field [| unTsMaybeInt       :: Maybe Int  |]
+        ],
+    Cons 'TsChar2Permuted [
+        Field [| unTsChar2PermutedR :: Char       |],
+        Field [| unTsChar2PermutedL :: Char       |]
+        ],
+    Cons 'TsPair [
+        Field [| unTsPairL          :: TestSimple |],
+        Field [| unTsPairR          :: TestSimple |]
+        ]
+    ]
+
+--------------------------------------------------------------------------------
+
+data TestIndexed
+    = TiInt Int
+    | TiIntList [Int]
+    | TiChar2 Char Char
+    | TiInteger Integer
+    | TiMaybeInt (Maybe Int)
+    | TiChar2Permuted Char Char
+    | TiPair TestIndexed TestIndexed
+    deriving (Eq, Show, Typeable)
+
```

Comment:
Bingo! This is exactly the sort of thing I was looking for!

Reply:
👍👍